collector/metrics.go

```diff
@@ -35,3 +38,99 @@ func calculateScheduledDurationPipelineRun(pipelineRun *v1beta1.PipelineRun) flo
 func calculateScheduledDurationTaskRun(taskrun *v1beta1.TaskRun) float64 {
 	return calcuateScheduledDuration(taskrun.CreationTimestamp.Time, taskrun.Status.StartTime.Time)
 }
+
+/*
+  Observed sequence from testing and tekton/k8s code examination:
+
+                  taskrun.create called by either user or pipelinerun reconciler, and k8s sets taskrun.create-time timestamp
+                         |
+                         |
+                  taskrun.start-time is set near top of reconcile loop, so this is bascially the first event into tekton after the create
+                         |
+                         |
+                  taskrun reonciler creates the pod and sets running condition in same loop (those 2 updates are processed by k8s/apiserver concurrently)
+                         |
+                        / \
+                       /   \
+ rcv run condition time     pod create time (no guarantee which is before, which is after)
+        |                                    |
+        |                                    |
+        |                   pod start-time set by kubelet when it "acknowledges" the pod, but has not pulled any container images
+        |                                    |
+        |                                    |
+        |                   pod container start times set by kubelet after images pulls and linux container start occurs
+        |
+ tekton pod reconciler
+ copies pod container
+ conditions to taskrun
+ step condition
+*/
+
+func calculateRunningConditionDurationTaskRun(taskrun *v1beta1.TaskRun, pod *corev1.Pod) float64 {
+	condition := taskrun.Status.GetCondition(apis.ConditionSucceeded)
+	if condition == nil {
+		return 0
+	}
+	transitionTime := condition.LastTransitionTime.Inner
+	if transitionTime.IsZero() || taskrun.CreationTimestamp.IsZero() {
+		return 0
+	}
+	// our filter should guarantee the pod is available, but we check for that just in case of intermittent hiccups with the api server etc.  Then,
+	// because of timing once the pod is created by tekton, and pod updates and taskrun updates occur in parallel,
+	// there can be variance on when various timestamps are set
+	if pod == nil || pod.Status.StartTime == nil || pod.Status.StartTime.IsZero() {
+		return transitionTime.Sub(taskrun.CreationTimestamp.Time).Seconds()
+
+	}
+	if transitionTime.Before(pod.Status.StartTime) {
+		// tekton has marked the taskrun as running before the kubelet has "accepted it",
+		// i.e. it set taskrun start time and the condition to running in the same reconcile loop
+		return 0
+	}
+	return transitionTime.Sub(pod.Status.StartTime.Time).Seconds()
```

Comment:
After some more testing with newer metrics, I'm thinking of redo-ing this, splitting up this metric a bit

Reply:
actually I've removed this metric .... confirmed in the end the pipeline controller was coupling even more field updates than I first thought